src/domain.jl

```diff
@@ -0,0 +1,174 @@
+    abstol::T
+    scalefactor::S
+    u::uType
+end
+
+struct GeneralDomainAffect{F,T,S,uType} <: AbstractDomainAffect{T,S,uType}
+    g::F
+    abstol::T
+    scalefactor::S
+    u::uType
+    resid::uType
+end
+
+# definitions of callback functions
+
+# Workaround since it is not possible to add methods to an abstract type:
+# https://github.com/JuliaLang/julia/issues/14919
+(f::PositiveDomainAffect)(integrator) = affect!(integrator, f)
+(f::GeneralDomainAffect)(integrator) = affect!(integrator, f)
+
+# general method defintions for domain callbacks
+
+"""
+    affect!(integrator, f::AbstractDomainAffect)
+
+Apply domain callback `f` to `integrator`.
+"""
+function affect!(integrator, f::AbstractDomainAffect{T,S,uType}) where {T,S,uType}
+    # modify u
+    u_modified!(integrator, modify_u!(integrator, f))
+
+    # define array of next time step, absolute tolerance, and scale factor
+    u = typeof(f.u) <: Void ? similar(integrator.u) : f.u
+    abstol = typeof(f.abstol) <: Void ? integrator.opts.abstol : f.abstol
+    scalefactor = typeof(f.scalefactor) <: Void ? integrator.opts.qmin : f.scalefactor
+
+    # setup callback and save addtional arguments for checking next time step
+    args = setup(f, integrator)
+
+    # cache current time step
+    dt = integrator.dt
+
+    # update time step of integrator to proposed next time step
+    integrator.dt = get_proposed_dt(integrator)
+
+    # adjust time step to bounds and time stops
+    fix_dt_at_bounds!(integrator)
+    modify_dt_for_tstops!(integrator)
+
+    while integrator.tdir * integrator.dt > 0
+        # calculate estimated value of next step and its residuals
+        integrator(u, integrator.t + integrator.dt)
+
+        # check whether time step is accepted
+        isaccepted(u, abstol, f, args...) && break
+
+        # adjust time step
+        dtcache = integrator.dt
+        integrator.dt *= scalefactor
+        fix_dt_at_bounds!(integrator)
+        modify_dt_for_tstops!(integrator)
+
+        # abort iteration when time step is not changed
+        if dtcache == integrator.dt
+            if integrator.opts.verbose
+                warn("Could not restrict values to domain. Iteration was canceled since ",
+                     "time step dt = ", integrator.dt, " could not be reduced.")
+            end
+            break
+        end
+    end
+
+    # update current and next time step
+    set_proposed_dt!(integrator, integrator.dt)
+    integrator.dt = dt
+end
+
+"""
+    modify_u!(integrator, f::AbstractDomainAffect)
+
+Modify current state vector `u` of `integrator` if required, and return whether it actually
+was modified.
+"""
+modify_u!(integrator, ::AbstractDomainAffect) = false
+
+"""
+    setup(f::AbstractDomainAffect, integrator)
+
+Setup callback `f` and return an arbitrary tuple whose elements are used as additional
+arguments in checking whether time step is accepted.
+"""
+setup(::AbstractDomainAffect, integrator) = ()
```

Comment:
Oh yes these methods seem strange... It was more a try to reduce duplicate code (since both algorithms are very similar), and I thought one could also extend the idea of domain callbacks and maybe come up with other algorithms. Currently they only would have to contain fields `abstol`, `u`, and `scalefactor`, so only those can be accessed inside the general default method `affect!`. However, e.g. the general domain callback has to calculate residuals in every iteration, and so by implementing this method we can setup a new array that can be returned inside this tuple and then will be used to check whether a calculated step is acceptable - this prevents reallocation of an array in every loop.

But I thought there should be a more elegant way to achieve this...